### PR TITLE
comp/trace: wrap pkg/trace/agent as a component

### PR DESCRIPTION
**Pull request by darccio**
### What does this PR do?

Introduce a wrapper for `pkg/trace/agent` as a component. Important things to know:

* Almost all original agent's code is still in `pkg/trace/agent`. Some bits have been moved mainly to `comp/trace/agent/agent.go`, to encapsulate them as part of the agent's lifecycle.
* The one big exception is `cmd/trace-agent/subcommands/run/run.go`. Now it lives inside the component, as it contains logic to start several non-components that the agent requires. They are potential next candidates to become components.

For future reference, the gist about creating components (at least wrappers like this one) is:

* You need to think of components like a restaurant, it has providers that user required by multiple parts of our component (internal code or other components) and it has three phases: setup, start & stop.
* Setup is fulfilled using a provider, a function encapsulated in a `fx.Module` and annotated with `fx.Provide`, and the values/components provided by other providers. The function must return a value matching the exported `Component` interface in your package. It can be an empty interface.
* Components can require other provided components and supplied values. Defined in the `dependencies` struct, that must embed `fx.In` type,the provider function must receive this struct as argument.
* Start is done through an unexported method `start` registered as a lifecycle hook. Beware of the provided context, as it has a timeout. Don't use it for any long-living code, and also you must not block inside this block.
* Stop is done in the same fashion as start, but the method is called `stop`. The same conditions apply regarding the provided context.
* If your component must handle signals to interrupt the component, use the injected `fx.Shutdowner` when catching a signal that means you need to stop.

### Additional Notes

* It was not possible to test the command line parsing as now we use `fxutil.Run` instead of `fxutil.OneShot`.
* Make sure you have [pyinvoke](https://www.pyinvoke.org/installing.html) installed. You need to run `inv lint-components --fix` before pushing, so you can avoid a failing CI.

### Possible Drawbacks / Trade-offs

* Introducing possible bugs due to unexpected changes in the processing of the configuration or the altered execution order of some parts (like CPU & memory profiling), or not using the provided context when running as Windows service.

### Describe how to test/QA your changes

- [x] Run it in testbed environment.
- [x] Dogfood it in local.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/trace/agent/agent.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/DataDog/datadog-agent/pkg/trace/telemetry"
+)
+
+type dependencies struct {
+	fx.In
+
+	Lc         fx.Lifecycle
+	Shutdowner fx.Shutdowner
+
+	Ctx                context.Context
```
Comment (hush-hush) [inline comp/trace/agent/agent.go:24]: Why do you need a context here? This feels like it should be handle by FX lifecycle for everything related to timeout and start/stop signal.

File: comp/trace/config/config_mock.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+	pkgconfig "github.com/DataDog/datadog-agent/pkg/config"
+)
+
+// newMock exported mock builder to allow modifying mocks that might be
+// supplied in tests and used for dep injection.
+func newMock(deps dependencies, t testing.TB) (Component, error) {
+	// injected Agent config should be a mock
+
+	traceCfg, err := setupConfig(deps, "apikey")
```
Comment (hush-hush) [inline comp/trace/config/config_mock.go:22]: Will the config be cleaned after the test is ran ?

Comment (darccio) [reply comp/trace/config/config_mock.go:22]: I'll check this.

File: comp/trace/agent/agent.go (hunk 1 above)
Comment (darccio) [reply comp/trace/agent/agent.go:24]: As it's a wrapper around pkg/trace, I didn't wanted to replace yet how stopping is handled by the tracer agent. It uses context cancellation to stop.

I'll take a second look to it, to make sure if it's viable to do in this first PR.

Comment (hush-hush) [reply comp/trace/agent/agent.go:24]: Your comp could create the context and cancel it when `Stop` is called by the FX lifecycle. Having a component depending on a context is a bit weird.

File: comp/trace/agent/agent.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	Config             config.Component
+	TelemetryCollector telemetry.TelemetryCollector
+}
+
+type agent struct {
+	*pkgagent.Agent
+
+	// TODO(AIT-8301): What is this used for?
+	shutdowner fx.Shutdowner
```
Comment (ogaca-dd) [inline comp/trace/agent/agent.go:33]: It allows to shutdown the Agent. As it is unused you can remove it.

File: comp/trace/agent/component.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
+
+	"go.uber.org/fx"
+)
+
+type Component interface {
+	// Run starts the tracer agent.
+	Run()
```
Comment (ogaca-dd) [inline comp/trace/agent/component.go:16]: Is it possible to not have a `Run` function? The component should be ready once built of use [lifecycle hook](https://uber-go.github.io/fx/lifecycle.html#lifecycle-hooks).

File: comp/trace/bundle_mock.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	coreconfig "github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/comp/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
+)
+
+// team: agent-apm
+
+// MockBundle defines the fx options for this bundle.
+var MockBundle = fxutil.Bundle(
```
Comment (ogaca-dd) [inline comp/trace/bundle_mock.go:27]: Bundles should contain only internal components. The reason is that if 2 bundles use `config.MockModule` then they cannot be used together.

File: comp/trace/bundle_test.go [hunk 6]
```diff
@@ -12,26 +12,25 @@ import (
...
 
-	coreconfig "github.com/DataDog/datadog-agent/comp/core/config"
-	"github.com/DataDog/datadog-agent/comp/trace/config"
+	"github.com/DataDog/datadog-agent/comp/core/config"
 	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
 )
 
 // team: agent-apm
 
+// TODO(AIT-8301): Why are these tests green if not all elements are supplied?
```
Comment (ogaca-dd) [inline comp/trace/bundle_test.go:21]: You have this issue because you have to explicitly instance your main components: https://github.com/DataDog/datadog-agent/pull/19466/files#diff-f25a1b5a7f28bd1a8890759fff9c2b2e0d0b8bbfc3ab1ba70bc4eb5aad6ef8bfL24-L27.

File: comp/trace/agent/agent.go (hunk 3 above)
Comment (darccio) [reply comp/trace/agent/agent.go:33]: In my next iteration I'm using it. I'll push it soon and mark it for code review. Thanks!

File: cmd/trace-agent/subcommands/run/command.go [hunk 7]
```diff
@@ -58,15 +57,34 @@ func runFx(ctx context.Context, cliParams *RunParams, defaultConfPath string) er
...
 	}
-	return fxutil.OneShot(Run,
-		fx.Supply(&contextSupplier{ctx: ctx}),
+	return fxutil.OneShot(Start,
+		fx.Supply(ctx),
 		fx.Supply(cliParams),
-		config.Module,
 		fx.Supply(coreconfig.NewAgentParamsWithSecrets(cliParams.ConfPath)),
+		// TODO(AIT-8301): Test this injection
+		fx.Provide(func(config config.Component) telemetry.TelemetryCollector {
```
[the remark this answers is not in the document]
Comment (darccio) [reply cmd/trace-agent/subcommands/run/command.go:65]: You are right, it was a half-cooked implementation and I thought I needed that but not really. I'll push my next iteration soon.

File: cmd/trace-agent/main.go [hunk 8]
```diff
@@ -12,10 +12,8 @@ import (
 	"github.com/DataDog/datadog-agent/cmd/trace-agent/command"
 )
 
-const defaultLogFile = "/var/log/datadog/trace-agent.log"
```
Comment (darccio) [inline cmd/trace-agent/main.go:15]: I removed this because it's duplicating `comp/trace/config.DefaultLogFilePath`.

File: pkg/trace/agent/agent.go [hunk 9]
```diff
@@ -180,7 +180,6 @@ func (a *Agent) work() {
 		}
 		a.Process(p)
 	}
-
```
Comment (darccio) [inline pkg/trace/agent/agent.go:183]: Housekeeping to move to a different PR.

File: comp/README.md [hunk 10]
```diff
@@ -49,8 +49,6 @@ Package telemetry implements a component for all agent telemetry.
 
 *Datadog Team*: agent-metrics-logs
 
-
-
```
Comment (darccio) [inline comp/README.md:53]: Housekeeping to move to a different PR.

File: cmd/trace-agent/command/command_test.go
Comment (darccio) [inline cmd/trace-agent/command/command_test.go:1]: Removed because I was unable to test the command parsing without running the tracer agent's `fx.App`.

**Comment by darccio** [comment]
Ready for review, but probably tests will fail. I'll fix them tomorrow.

File: comp/trace/agent/agent.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+type agent struct {
+	*pkgagent.Agent
+
+	cancel             context.CancelFunc
+	config             config.Component
+	ctx                context.Context
+	params             *Params
+	shutter            fx.Shutdowner
```
Comment (katiehockman) [inline comp/trace/agent/agent.go:43]:
```suggestion
	shutdowner            fx.Shutdowner
```

File: comp/trace/agent/agent.go
Comment (katiehockman) [inline comp/trace/agent/agent.go:1]: nit (optional): I suggest reducing the newlines in the functions. Don't need one between every if-statement, as it is now. I'd just suggest putting them structurally where a major context switch needs to happen.

File: comp/trace/agent/agent.go [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+// handleSignal closes a channel to exit cleanly from routines
+func handleSignal(shutter fx.Shutdowner) {
+	sigChan := make(chan os.Signal, 10)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGPIPE)
+	for signo := range sigChan {
+		switch signo {
+		case syscall.SIGINT, syscall.SIGTERM:
+			log.Infof("received signal %d (%v)", signo, signo)
+			_ = shutter.Shutdown()
```
Comment (katiehockman) [inline comp/trace/agent/agent.go:144]: Why not just `shutter.Shutdown()` without the `_ =`?

same question for the `log.Warnf` line

File: comp/trace/agent/agent.go [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+		}
+		f.Close()
+	}
+
+	return nil
+}
+
+// handleSignal closes a channel to exit cleanly from routines
+func handleSignal(shutter fx.Shutdowner) {
+	sigChan := make(chan os.Signal, 10)
```
Comment (katiehockman) [inline comp/trace/agent/agent.go:138]: why `10`?

File: comp/trace/agent/agent.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGPIPE)
+	for signo := range sigChan {
+		switch signo {
+		case syscall.SIGINT, syscall.SIGTERM:
+			log.Infof("received signal %d (%v)", signo, signo)
```
Comment (katiehockman) [inline comp/trace/agent/agent.go:143]:
```suggestion
			log.Infof("Received signal %d (%v). Shutting down.", signo, signo)
```

File: comp/trace/agent/agent.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+			_ = shutter.Shutdown()
+			return
+		case syscall.SIGPIPE:
+			// By default systemd redirects the stdout to journald. When journald is stopped or crashes we receive a SIGPIPE signal.
+			// Go ignores SIGPIPE signals unless it is when stdout or stdout is closed, in this case the agent is stopped.
+			// We never want the agent to stop upon receiving SIGPIPE, so we intercept the SIGPIPE signals and just discard them.
+		default:
+			_ = log.Warnf("unhandled signal %d (%v)", signo, signo)
```
Comment (katiehockman) [inline comp/trace/agent/agent.go:151]:
```suggestion
			_ = log.Warnf("Unhandled signal %d (%v)", signo, signo)
```

**Review by katiehockman — COMMENTED** [review]
Overall, looks good. Just a few minor comments.

Can you add more details to the PR description, in addition to filling out the sections in the existing template? I feel that using the components framework is pretty nuanced, and it would be very helpful to have more details around what decision decisions were made here, what shared components the trace-agent is now using, which it isn't, future work planned, etc.

File: comp/trace/agent/agent.go (hunk 12 above)
Comment (darccio) [reply comp/trace/agent/agent.go:144]: @katiehockman I think I have a different set of linters in my local golangci-lint, and one of them is the unchecked error linter. These lines were just to make sure the unchecked error is explicitly unchecked.

I'll compare CI and local configurations for golangci-lint, to make sure I follow the same conventions.

File: comp/trace/agent/agent.go (hunk 13 above)
Comment (darccio) [reply comp/trace/agent/agent.go:138]: @katiehockman That was actually like this in the original code that I moved from `pkg/trace/agent`. I left it as is.

File: comp/trace/agent/agent.go (hunk 11 above)
Comment (darccio) [reply comp/trace/agent/agent.go:43]: This is also a change to make the IDE happy. Goland has a grammar checker that I don't find how to fully deactivate, but I'll try that again.

File: comp/trace/agent/agent.go
Comment (darccio) [reply comp/trace/agent/agent.go:1]: It was like this in the original code too 😄 There is a linter that I like that sets a proper block structure: https://golangci-lint.run/usage/linters/#wsl 

Anyway, I'll fix them.

File: comp/trace/agent/agent.go (hunk 13 above)
Comment (darccio) [reply comp/trace/agent/agent.go:138]: Curious: almost everywhere we use buffered channels with 1 element length, except here and some code in `pkg/proces/util` (where the channel has a length of 100 elements).

According to `os/signal` docs:

> Package signal will not block sending to c: the caller must ensure that c has sufficient buffer space to keep up with the expected signal rate. For a channel used for notification of just one signal value, a buffer of size 1 is sufficient.

There seems not to be a good reason to create a buffered channel with length bigger than 1, so I fixed it.

File: comp/trace/agent/agent.go (hunk 12 above)
Comment (darccio) [reply comp/trace/agent/agent.go]: The Windows linter seems to require comply with `errcheck` linter: https://github.com/DataDog/datadog-agent/actions/runs/6249078120/job/16964969083?pr=19466#step:7:210

To get around it, I'm going to settle with the underscore explicit syntax.

File: cmd/trace-agent/subcommands/run/command.go [hunk 16]
```diff
@@ -49,40 +41,25 @@ func setParamFlags(cmd *cobra.Command, cliParams *RunParams) {
...
 		fx.Supply(coreconfig.NewAgentParamsWithSecrets(cliParams.ConfPath)),
 		coreconfig.Module,
+		config.Module,
+		// Required to avoid cyclic imports.
+		fx.Supply(agent.Params{
+			CPUProfile:  cliParams.CPUProfile,
+			MemProfile:  cliParams.MemProfile,
+			PIDFilePath: cliParams.PIDFilePath,
+		}),
+		trace.Bundle,
```
Comment (darccio) [inline cmd/trace-agent/subcommands/run/command.go:63]: @truthbk @hush-hush @ogaca-dd After some debugging, I found out that my component's `newAgent` and `start` method are never called.

My guess is that I'm not doing properly the injection here. Right now, when the agent starts it blocks on `fxutil.Run` waiting on `app.Done()`.

That's why it doesn't show any logs, causing `TestCreditCards` to fail because it looks for a specific text in the agent logs and timeouts after 10 seconds. This test has no relation with the agent itself, but it's part of the integration tests. They build the trace-agent on the fly and run the binary, so they are acting as smoke tests in the current situation.

I'd appreciate your feedback. Thanks!

File: comp/trace/agent/agent.go (hunk 13 above)
Comment (katiehockman) [reply comp/trace/agent/agent.go]: 🙏  Thanks

**Review by katiehockman — APPROVED** [review]

File: cmd/trace-agent/subcommands/run/command.go [hunk 17]
```diff
@@ -49,40 +41,25 @@ func setParamFlags(cmd *cobra.Command, cliParams *RunParams) {
...
-func runFx(ctx context.Context, cliParams *RunParams, defaultConfPath string) error {
+func runFx(cliParams *RunParams, defaultConfPath string) error {
 	if cliParams.ConfPath == "" {
 		cliParams.ConfPath = defaultConfPath
 	}
-	return fxutil.OneShot(Run,
-		fx.Supply(&contextSupplier{ctx: ctx}),
-		fx.Supply(cliParams),
-		config.Module,
+	return fxutil.Run(
```
Comment (ogaca-dd) [inline cmd/trace-agent/subcommands/run/command.go:53]: Is it expected to not call `Run` function (`fxutil.OneShot(Run`)?  Can you explain why you changed from `fxutil.OneShot` to ` fxutil.Run`?

File: cmd/trace-agent/subcommands/run/command.go (hunk 16 above)
Comment (ogaca-dd) [reply cmd/trace-agent/subcommands/run/command.go:63]: Your component is created only if it is used. As your component is defined as an [empty interface](https://github.com/DataDog/datadog-agent/pull/19466/files#diff-6699922c80e372c638ad4655da617024d3e81f35f360d876f6901cc208f949eeR17) you have to force the creation of your component. You can do that by adding [here](https://github.com/DataDog/datadog-agent/pull/19466/files#diff-6699922c80e372c638ad4655da617024d3e81f35f360d876f6901cc208f949eeR21)  something like `fx.Invoke(_  agent.Component) {}`

File: comp/trace/bundle_test.go [hunk 18]
```diff
@@ -23,29 +24,25 @@ func TestBundleDependencies(t *testing.T) {
 	require.NoError(t, fx.ValidateApp(
 		// instantiate all of the core components, since this is not done
 		// automatically.
-		fx.Invoke(func(r config.Component) {}),
-		fx.Invoke(func(coreconfig.Component) {}),
```
Comment (ogaca-dd) [inline comp/trace/bundle_test.go:27]: You should instanciate all your components to make the test useful.

File: cmd/trace-agent/subcommands/run/command.go (hunk 17 above)
Comment (darccio) [reply cmd/trace-agent/subcommands/run/command.go:53]: @ogaca-dd I expected to be more idiomatic - in `fx` ecosystem - to encapsulate everything in the component, and rely in `fx.Run` to initialize everything.

Comment (ogaca-dd) [reply cmd/trace-agent/subcommands/run/command.go:53]: `fxutil.Run` is used only in [systray](https://github.com/DataDog/datadog-agent/blob/972c3b86fd2bffbf615b31b55db2f5fffba6879a/cmd/systray/command/command.go#L82) and unclear why `fxutil.OneShot` is not used. `fxutil.OneShot` is used at more than 100 locations. So except if there is a specific reason to use `fx.Run`, I think be consistent with the rest of the code is important. What do you think?

Comment (darccio) [reply cmd/trace-agent/subcommands/run/command.go:53]: @ogaca-dd Sure, I prefer to be consistent. I'll rework it to use `fxutil.OneShot` again.

File: comp/trace/bundle_test.go (hunk 18 above)
Comment (darccio) [reply comp/trace/bundle_test.go:27]: Fixed 😄 I finally understood `fx.Invoke`.

File: cmd/trace-agent/subcommands/run/command.go (hunk 17 above)
Comment (ogaca-dd) [reply cmd/trace-agent/subcommands/run/command.go:53]: Sorry, I had another look and it is fine to use `fxutil.Run`.

File: comp/trace/bundle_test.go [hunk 19]
```diff
@@ -23,11 +23,15 @@ import (
 func TestBundleDependencies(t *testing.T) {
 	require.NoError(t, fx.ValidateApp(
 		// instantiate all of the core components, since this is not done
-		// automatically.
-		coreconfig.Module,
+		// automatically. Use fx.Invoke to make sure components are initialized
+		// and all the providers are called.
 		fx.Supply(coreconfig.Params{}),
+		fx.Invoke(func(_ coreconfig.Component) {}),
```
Comment (ogaca-dd) [inline comp/trace/bundle_test.go:29]: You don't have to call coreconfig.Component as it is not part of your Bundle.

File: comp/trace/bundle_test.go [hunk 20]
```diff
@@ -23,11 +23,15 @@ import (
...
+		coreconfig.Module,
+		fx.Invoke(func(_ config.Component) {}),
```
Comment (ogaca-dd) [inline comp/trace/bundle_test.go:31]: Is it expected `Bundle` doesn't contain `config.Module`?

File: comp/trace/bundle_test.go [hunk 21]
```diff
@@ -38,14 +42,16 @@ func TestMockBundleDependencies(t *testing.T) {
 	os.Setenv("DD_DD_URL", "https://example.com")
 	defer func() { os.Unsetenv("DD_DD_URL") }()
 
-	config := fxutil.Test[config.Component](t, fx.Options(
-		coreconfig.MockModule,
+	cfg := fxutil.Test[config.Component](t, fx.Options(
 		fx.Supply(coreconfig.Params{}),
+		fx.Invoke(func(_ coreconfig.Component) {}),
```
Comment (ogaca-dd) [inline comp/trace/bundle_test.go:47]: core config is not part of `MockBundle`.

**Review by ogaca-dd — APPROVED** [review]

File: comp/trace/bundle_test.go (hunk 20 above)
Comment (darccio) [reply comp/trace/bundle_test.go:31]: @ogaca-dd I checked the original code from @truthbk and it contained it. Probably it went missing during all the tests changing the code to fit in it.

Comment (darccio) [reply comp/trace/bundle_test.go:30]: Anyway, I updated the PR with this and all the comments. Please check.

**Review by ogaca-dd — APPROVED** [review]

**Review by truthbk — APPROVED** [review]

**Review by hush-hush — APPROVED** [review]

**Review by ogaca-dd — CHANGES_REQUESTED** [review]
LGTM but I think not returning a nil component is important.

File: cmd/trace-agent/command/command_test.go [hunk 22]
```diff
@@ -1,27 +0,0 @@
...
-	"testing"
-
-	"github.com/DataDog/datadog-agent/cmd/trace-agent/subcommands/run"
-	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
-	"github.com/spf13/cobra"
-	"github.com/stretchr/testify/require"
-)
-
-func TestRootCommand(t *testing.T) {
-	fxutil.TestOneShotSubcommand(t,
```
Comment (ogaca-dd) [inline cmd/trace-agent/command/command_test.go:18]: I think it is important to have a unit test that checks there is no missing dependencies. 
You can create easily a test by having a function that returns [all your fx.Option](https://github.com/DataDog/datadog-agent/pull/19466/files#diff-997cf98dbac166c4dfac8749608990bb76203def6b7a03a3a7e6efe05176d653R67-R97) and check there is no missing dependencies using [fx.ValidateApp](https://pkg.go.dev/go.uber.org/fx#ValidateApp)

File: cmd/trace-agent/subcommands/run/command_test.go [hunk 23]
```diff
@@ -1,33 +0,0 @@
...
-	"testing"
-
-	"github.com/DataDog/datadog-agent/cmd/trace-agent/subcommands"
-	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
-	"github.com/spf13/cobra"
-	"github.com/stretchr/testify/require"
-)
-
-func TestStartCommand(t *testing.T) {
-	fxutil.TestOneShotSubcommand(t,
```
Comment (ogaca-dd) [inline cmd/trace-agent/subcommands/run/command_test.go:18]: Same comment as for cmd/trace-agent/command/command_test.go

File: cmd/trace-agent/subcommands/run/command.go [hunk 24]
```diff
@@ -66,6 +76,83 @@ func runFx(ctx context.Context, cliParams *RunParams, defaultConfPath string) er
 			PIDFilePath: cliParams.PIDFilePath,
 		}),
 		trace.Bundle,
+		// TODO: corelogger must be a component (for future reference)
+		fx.Invoke(func(cfg config.Component, telemetryCollector telemetry.TelemetryCollector) error {
+			tracecfg := cfg.Object()
+			if err := pkgconfig.SetupLogger(
```
[the remark this answers is not in the document]
Comment (darccio) [reply cmd/trace-agent/subcommands/run/command.go:82]: @ogaca-dd I need to initialize the logger as early as possible. Manual QA shown that log statements were missing comparing to agent compiled from main.

File: comp/trace/agent/agent.go [hunk 25]
```diff
@@ -53,38 +57,37 @@ type agent struct {
...
 }
 
-func newAgent(deps dependencies) Component {
-	telemetryCollector := telemetry.NewCollector(deps.Config.Object())
+func newAgent(deps dependencies) (Component, error) {
+	tracecfg := deps.Config.Object()
+	if !tracecfg.Enabled {
+		log.Info(messageAgentDisabled)
+		deps.TelemetryCollector.SendStartupError(telemetry.TraceAgentNotEnabled, fmt.Errorf(""))
+		return nil, ErrAgentDisabled
```
[the remark this answers is not in the document]
Comment (darccio) [reply comp/trace/agent/agent.go:66]: I'll check, but I need to abort the creation of the component there. Fx seems to just handle fine this as long as we return an error.

File: cmd/trace-agent/subcommands/run/command.go [hunk 26]
```diff
@@ -46,11 +48,14 @@ func setParamFlags(cmd *cobra.Command, cliParams *RunParams) {
...
 
-func runFx(cliParams *RunParams, defaultConfPath string) error {
+func runFx(ctx context.Context, cliParams *RunParams, defaultConfPath string) error {
 	if cliParams.ConfPath == "" {
 		cliParams.ConfPath = defaultConfPath
 	}
 	return fxutil.Run(
+		// ctx is required to be supplied from here, as Windows needs to inject its own context
+		// to allow the agent to work as a service.
+		fx.Provide(func() context.Context { return ctx }),
```
[the remark this answers is not in the document]
Comment (darccio) [reply cmd/trace-agent/subcommands/run/command.go]: I tried, but unfortunately it fails with this error:

```
missing type:
	- context.Context (did you mean to Provide it?)
```

As we won't be focusing on components this next quarter, I'll leave it with `fx.Provide` as it worked as expected. We'll revisit this in the future.

File: cmd/trace-agent/command/command_test.go (hunk 22 above)
Comment (darccio) [reply cmd/trace-agent/command/command_test.go:18]: If I understand correctly, that's what `TestBundleDependencies` and `TestMockBundleDependencies` do in `comp/trace/bundle_test.go`.

File: comp/trace/agent/agent.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (darccio) [reply comp/trace/agent/agent.go:66]: Done.

File: comp/trace/agent/agent.go [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+	ctx                context.Context
+	params             *Params
+	shutdowner         fx.Shutdowner
+	telemetryCollector telemetry.TelemetryCollector
+	wg                 sync.WaitGroup
+}
+
+func newAgent(deps dependencies) Component {
+	c := component{}
+	deps.Lc.Append(fx.Hook{OnStart: c.start, OnStop: c.stop})
```
Comment (ogaca-dd) [inline comp/trace/agent/agent.go:64]: Usually `deps.Lc.Append` is called after the check if the component is enabled.
This would make the code easier because you can capture the instance of `agent` and so don't need to have `Optional` field.
The following code
```go
deps.Lc.Append(fx.Hook{OnStart: c.start, OnStop: c.stop})
tracecfg := deps.Config.Object()
	if !tracecfg.Enabled {
           // .... 
        }
// ....
c.Optional = util.NewOptional[agentComponent](ag)
```
can be replaced by
```go
tracecfg := deps.Config.Object()
	if !tracecfg.Enabled {
           // .... 
        }
// ... 
deps.Lc.Append(fx.Hook{
      OnStart: func(_ context.Context) error { return start(ag)}),
       OnStop: func(_ context.Context) error { return stop(ag)) })
```

**Review by ogaca-dd — APPROVED** [review]

File: comp/trace/agent/agent.go [hunk 28]
```diff
@@ -85,20 +85,17 @@ func newAgent(deps dependencies) Component {
 		telemetryCollector: deps.TelemetryCollector,
 		wg:                 sync.WaitGroup{},
 	}
-	c.Optional = util.NewOptional[agentComponent](ag)
+	deps.Lc.Append(fx.Hook{
+		// Provided contexts have a timeout, so it can't be used for gracefully stopping long-running components.
+		// These contexts are cancelled on a deadline, so they would have side effects on the agent.
+		OnStart: func(_ context.Context) error { return start(ag) },
+		OnStop:  func(_ context.Context) error { return stop(ag) }})
+	c.agent = ag
```
Comment (ogaca-dd) [inline comp/trace/agent/agent.go:95]: Note: This assignment is not needed.

**Review by ogaca-dd — APPROVED** [review]

**Review by chouquette — APPROVED** [review]
LGTM for agent-platform
